proof/crefine/AARCH64/Fastpath_C.thy

```diff
@@ -552,271 +541,236 @@ lemma ccorres_catch_bindE_symb_exec_l:
+  assumes kos: "\<And>ko p. f \<lbrace>\<lambda>s. ko_at' (ko::asidpool) p s \<rbrace>"
+  shows "f \<lbrace> asid_has_vmid asid vmid vsroot \<rbrace>"
+  unfolding asid_has_vmid_def asid_has_entry_def
+  apply (subst case_option_If2)+
+  apply wp_pre
+   apply (wps asids)
+   apply (rule hoare_vcg_if_lift) (* does not work when combined with wpsimp *)
+   apply (wpsimp wp: hoare_vcg_imp_lift' hoare_vcg_ex_lift hoare_vcg_conj_lift[OF kos])
+  apply auto
   done
 
-crunches vcpuSwitch
-  for pd_at_asid'[wp]: "pd_at_asid' pd asid"
-  and pd_has_hwasid[wp]: "pd_has_hwasid pd"
-  (ignore: modifyArchState wp: crunch_wps simp: crunch_simps)
+lemma vcpuSwitch_asid_has_vmid[wp]:
+  "vcpuSwitch t \<lbrace>asid_has_vmid asid vmid vsroot\<rbrace>"
+  unfolding vcpuSwitch_def
+  by (wpc
+      | wpsimp wp: asid_has_vmid_lift modifyArchState_valid_lift simp: split_def
+      | simp add: modifyArchState_def)+
 
+(* FIXME AARCH64 for the 0x20, see valid_cnode_cap_cte_at'' *)
 lemma switchToThread_fp_ccorres:
-  "ccorres dc xfdc (pspace_aligned' and pspace_distinct' and valid_objs' and no_0_obj'
-                          and valid_pde_mappings' and valid_arch_state'
-                          and tcb_at' thread
-                          and cte_wp_at' (\<lambda>cte. isValidVTableRoot (cteCap cte)
-                                              \<and> capPDBasePtr (capCap (cteCap cte)) = pd)
-                                         (thread + tcbVTableSlot * 0x10)
-                          and pd_has_hwasid pd
-                          and (\<lambda>s. asid_map_pd_to_hwasids (armKSASIDMap (ksArchState s)) pd
-                                     = set_option (pde_stored_asid v)))
-                   (UNIV \<inter> {s. thread_' s = tcb_ptr_to_ctcb_ptr thread}
-                         \<inter> {s. cap_pd_' s = pde_Ptr pd}
-                         \<inter> {s. stored_hw_asid___struct_pde_C_' s = v}) []
-      (Arch.switchToThread thread
-            >>= (\<lambda>_. setCurThread thread))
-      (Call switchToThread_fp_'proc)"
-  apply (cinit' lift: thread_' cap_pd_' stored_hw_asid___struct_pde_C_')
-   apply (simp add: ARM_HYP_H.switchToThread_def bind_assoc setVMRoot_def
-                    cap_case_isPageDirectoryCap
-               del: Collect_const cong: call_ignore_cong)
-   apply (simp add: getThreadVSpaceRoot_def locateSlot_conv getSlotCap_def
-               del: Collect_const cong: call_ignore_cong)
-   apply simp
+  "ccorres dc xfdc
+     (pspace_aligned' and pspace_distinct' and valid_objs' and no_0_obj' and valid_arch_state'
+      and tcb_at' thread
+      and cte_wp_at' (\<lambda>cte. isValidVTableRoot (cteCap cte)
+                            \<and> capPTBasePtr (capCap (cteCap cte)) = vsroot
+                            \<and> (\<exists>vaddr. capPTMappedAddress (capCap (cteCap cte))
+                                       = Some (asid, vaddr)))
+                     (thread + tcbVTableSlot * 0x20)
```

Comment:
Is this the `cteSize` or something that you were asking about on Mattermost? Is there any benefit of using that instead of the bare number here?

Reply:
Yeah, this is the part that refers to maybe fixing it:
```
(* FIXME AARCH64 in this file, we see ptr + 0x20 * tcbVTableSlot a lot, where the 0x20 is
   2 ^ cte_level_bits
   This used to be 0x10 on 32-bit, so unlikely to change soon, but still might be worth a cleanup
   or abbreviation. This is one way it could work (and can go outside Arch locale): *)
declare cte_level_bits_def[code]
value_abbreviation (input) cte_size "(2::machine_word) ^ cte_level_bits"
```

It's not really clear if an input abbreviation only will be helpful here. Might be better to go with a constant, but it was hard to wrangle it in the proofs. Still not sure about this one, but yeah it's what I was asking about in mattermost